stonkers/__init__.py

```diff
@@ -1,108 +1,188 @@
 import yaml
+from collections import Counter
 import pandas as pd
 import numpy as np
 from yahoo_fin import stock_info as si
 
 
 class stock_portfolio:
-    """Stock Portfolio Class
-    """
+    """Stock Portfolio Class"""
 
     def __init__(self, portfolio_filename=None):
         if portfolio_filename:
             with open(portfolio_filename) as f:
                 yaml_objects = yaml.safe_load(f)
-                self.holding_percentage = yaml_objects["port_goals"]
+                self.goal_percentage = yaml_objects["port_goals"]
                 self.holdings = yaml_objects["port_holds"]
         else:
-            self.holding_percentage = {}
+            self.goal_percentage = {}
             self.holdings = {}
         self.cols = [
             "ticker",
             "goal_percentage",
             "holding_percentage",
             "holdings",
             "stock_price",
-            "holdings_value",
-            "goal_holding_value",
             "holding_value_diff",
+            "need_to_buy_or_sell",
+            "need_to_buy",
         ]
 
     def add_holding(self, ticker, amount):
         assert (type(ticker)) == str
         assert (type(amount)) == int
         self.holdings[ticker] = amount
 
-    def add_percent_holding(self, ticker, percentage):
+    def add_goal_percent(self, ticker, percentage):
         assert (type(ticker)) == str
-        self.holding_percentage[ticker] = percentage
+        self.goal_percentage[ticker] = percentage
+
+    def check_percent_holdings(self, balance_percents=False):
+        percentage_sum = sum(self.goal_percentage.values())
+        if percentage_sum != 1:
+            print("Portfolio percentages do not sum to 100%")
+            if balance_percents:
+                for ticker in self.goal_percentage.keys():
+                    print("Rebalancing holding percentages to sum to 100%")
+                    self.goal_percentage[ticker] = (
+                        self.goal_percentage[ticker] / percentage_sum
+                    )
+
+    def pull_stock_prices(self):
+        self.stock_price = {}
+        for ticker in self.goal_percentage.keys():
+            self.stock_price[ticker] = si.get_live_price(ticker)
+
+    def calculate_portfolio_stats(self):
+        if not hasattr(self, "holdings"):
+            print("Please add hodlings before calculating stats")
+
+        elif not hasattr(self, "goal_percentage"):
+            print("Please add goal percentage before calculating stats")
+
+        else:
+            if not hasattr(self, "stock_price"):
+                self.pull_stock_prices()
+
+            self.holdings_value = {}
```

Comment:
Can probably do the same thing with all this logic here of setting these via dictionary comprehensions rather than having to set to empty and then iterate and set all individual keys.